### ISSUE TITLE
Use Visitor mechanism instead of Lint [phab:D987]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1319]
_Created by **djrenren** at 2016-07-28 00:30:41_: 

This change refactors the rust indexer to use Visitor to power its indexing in place of LateLintPass. This allows us to control the visiting behavior and avoid things like desugared for loop blocks or macro expansions.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-28 19:36:29_:

nit: should these `use` directives be sorted?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-28 19:36:29_:

nit: `VName` or `VName.`; use periods (or don't) throughout to end complete sentences in documentation.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-28 19:36:29_:

Why does this have to happen?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-28 19:36:29_:

nit: handled (and maybe a period).

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-07-28 21:01:29_:

  - Cleaned up comments and reordered use blocks

rustfmt doesn't do anything about the {..} variants in use statements but I ordered them just cuz. Comments don't use trailing periods anymore

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-28 21:05:52_:

The ordering I'd been thinking of was
```
use rustc::hir::def_id::DefId;
use rustc::hir::MatchSource::ForLoopDesugar;
```
vs the current ordering
```
use rustc::hir::MatchSource::ForLoopDesugar;
use rustc::hir::def_id::DefId;
```
but if this isn't actually something Rust programmers are concerned about, feel free to do nothing.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2016-07-28 21:05:52_

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-07-28 23:23:28_:

  - Reordered imports again

Totally should be kept in order. My bad.

**Comment by kythe** [comment]
_**djrenren** closed via commit a6a9aba531843360c7ba1998cf3293d2441ec781 at 2016-07-28 23:24:12_
